src/main/java/org/truffleruby/core/objectspace/ObjectSpaceNodes.java

```diff
@@ -98,38 +102,38 @@ protected boolean isFloatID(DynamicObject id) {
 
     }
 
-    @CoreMethod(names = "each_object", isModuleFunction = true, needsBlock = true, optional = 1, returnsEnumeratorIfNoBlock = true)
+    @CoreMethod(names = "all_objects", isModuleFunction = true, optional = 1)
     public abstract static class EachObjectNode extends YieldingCoreMethodNode {
 
         @TruffleBoundary // for the iterator
-        @Specialization
-        public int eachObject(NotProvided ofClass, DynamicObject block) {
-            int count = 0;
+        @Specialization(guards = "isNil(ofClass)")
+        public DynamicObject eachObject(DynamicObject ofClass) {
+            Set<DynamicObject> objects = ObjectGraph.stopAndGetAllObjects(this, getContext());
+            ArrayList<DynamicObject> objectList = new ArrayList<>(objects.size());
 
-            for (DynamicObject object : ObjectGraph.stopAndGetAllObjects(this, getContext())) {
+            for (DynamicObject object : objects) {
                 if (!isHidden(object)) {
-                    yield(block, object);
-                    count++;
+                    objectList.add(object);
                 }
             }
 
-            return count;
+            return createArray(objectList.toArray(), objectList.size());
```

Comment:
This is very significant overhead: it's a second list of all DynamicObject in the heap!